raft/server.cc

```diff
@@ -891,6 +871,16 @@ future<> server_impl::io_fiber(index_t last_stable) {
 
             // Process committed entries.
             if (batch.committed.size()) {
+                if (_conf_commit_promise) {
+                    for (const auto& e: batch.committed) {
+                        const auto* cfg = get_if<raft::configuration>(&e->data);
+                        if (cfg != nullptr && !cfg->is_joint()) {
```

Comment:
I think this problem could be easily fixed by just waiting on the joint entry before waiting on `_conf_commit_promise`.
If we ensure that the joint entry is committed, then the next non-joint entry following it must be for our config change, irrespective of whether there were any leader changes in the meantime.

Reply:
I tried this
```
auto f = _conf_commit_promise.emplace().get_future();
co_await wait_for_entry({.term = e.term, .idx = e.idx}, wait_type::committed, as);
co_await std::move(f);
```
and some nemesis tests break for some reason.

Current version seems to work:
1. we reset `_conf_commit_promise` as soon as we discover a lost of leadership
2. before that, if we managed to get committed non-joint config from fsm, it couldn't be anything else but our config:
    * we were the leader in the recent past and we posted new config
    * for other server to be become a leader and to be able to post a new config it would take several rounds of `io_fiber` to communicate `request_votes` between the servers, and at that time we would have already reset _conf_commit_promise.

Reply:
figured it out, the problem was that nobody handled an exception in _conf_commit_promise future, if an exception has also occured in wait_for_entry